pkg/set.go

```diff
@@ -0,0 +1,83 @@
+package adt
+
+var keyExists = struct{}{}
+
+type Set[E comparable] map[E]struct{}
+
+func NewSet[E comparable](xs ...E) Set[E] {
+	s := make(Set[E])
+	for _, x := range xs {
+		s[x] = keyExists
+	}
+	return s
+}
+
+func (s Set[E]) Add(xs ...E) {
+	for _, x := range xs {
+		s[x] = keyExists
+	}
+}
+
+func (s Set[E]) Remove(xs ...E) {
+	for _, x := range xs {
+		delete(s, x)
+	}
+}
+
+func (s Set[E]) Len() int {
+	return len(s)
+}
+
+func (s Set[E]) Contains(x E) bool {
+	_, exists := s[x]
+	return exists
+}
+
+func (s Set[E]) Equals(other Set[E]) bool {
+	if s.Len() != other.Len() {
+		return false
+	}
+
+	for x := range s {
+		if !other.Contains(x) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s Set[E]) Items() []E {
+	res := make([]E, 0, s.Len())
+	for x := range s {
+		res = append(res, x)
+	}
+	return res
+}
+
+func (s Set[E]) Clone() Set[E] {
+	return NewSet(s.Items()...)
+}
+
+func (s Set[E]) Difference(other Set[E]) Set[E] {
+	res := s.Clone()
```

Comment:
More waste: building a map and removing is slower that doing filtering in a for loop before inserting.

Reply:
fixed.